### DISCUSSION TITLE
How to use uri! macro with endpoint defined in another module?

### DISCUSSION THREAD

**Discussion by scullionw** [rwf2/Rocket#2720] [Questions]
when i import the endpoint handler i can assign it in `routes!`, but it doesn't work in the `uri!` macro

```rust
#[macro_use]
extern crate rocket;

#[tokio::main]
async fn main() {}

mod routes {
    pub mod deeper {
        #[get("/<name>")]
        pub fn hello(name: String) -> String {
            name
        }
    }
}

// #[cfg(test)]
mod tests {
    use super::*;
    use rocket::http::Status;
    use rocket::local::blocking::Client;
    use rocket::Build;
    use routes::deeper::hello;

    fn rocket() -> rocket::Rocket<Build> {
        rocket::build().mount("/", routes![hello]) // <--- works here
    }

    #[test]
    fn test_hello() {
        let client = Client::tracked(rocket()).expect("valid rocket instance");
        let response = client.get(uri!(hello(name = "world"))).dispatch(); // <---- this fails
        assert_eq!(response.status(), Status::Ok);
        assert_eq!(response.into_string(), Some("world".into()));
    }
}
```

```
error: cannot find macro `rocket_uri_macro_hello` in this scope
  --> src/main.rs:30:40
   |
30 |         let response = client.get(uri!(hello(name = "world"))).dispatch();
   |                                        ^^^^^
   |
help: consider importing this macro through its public re-export
   |
16 +     use crate::routes::rocket_uri_macro_hello;
   |

error: could not compile `sostreamy` (bin "sostreamy" test) due to 1 previous error
```

i also could not get it to work by adding `#[macro_use]`

it does work when using a module path in the name like routes::deeper::hello, but im trying to understand why and if it is possible to just import the name, like we can for assigning the routes.

edit: i do see the `rocket_uri_macro_hello` but I didn't see this naming anywhere in docs so i feel like there is something else i am missing

**Comment by SergioBenitez** [comment]
> it does work when using a module path in the name like routes::deeper::hello, but im trying to understand why and if it is possible to just import the name, like we can for assigning the routes.

That's the way to do it. This is covered in the docs. It is not possible import and then use it directly due to limitations in rust's proc macro infrastructure.
